File: 2744-find-maximum-number-of-string-pairs/2744-find-maximum-number-of-string-pairs.java
class Solution {

    private String reverseString(String str) {
        StringBuilder result = new StringBuilder();
        for (int i = str.length() - 1; i >= 0; i--) result.append(str.charAt(i));
        return result.toString();
    }

    public int maximumNumberOfStringPairs(String[] words) {
        int result = 0;

        for (int i = 0; i < words.length - 1; i++) {
            final String word1 = words[i];
            for (int j = i + 1; j < words.length; j++) {
                final String word2 = words[j];
                final String reversedWord2 = reverseString(word2);

                if (word1.equals(reversedWord2)) {
                    result += 1;
                }
            }
        }

        return result;
    }
}

File: 2744-find-maximum-number-of-string-pairs/2744-find-maximum-number-of-string-pairs.rs
use std::collections::HashSet;

impl Solution {
    pub fn maximum_number_of_string_pairs(words: Vec<String>) -> i32 {
        let mut result = 0;
        let mut used_index_set = HashSet::<usize>::new();

        for (i, word) in words.iter().enumerate() {
            if i == words.len() - 1 {
                break;
            }
            if used_index_set.contains(&i) {
                continue;
            }

            let reversed_string = word.chars().into_iter().rev().collect::<String>();

            for j in i + 1..words.len() {
                if used_index_set.contains(&j) {
                    continue;
                }
                let element = words.get(j).unwrap();

                if reversed_string == *element {
                    result += 1;
                    used_index_set.insert(i);
                    used_index_set.insert(j);
                    break;
                }
            }
        }

        result
    }
}
